### PR TITLE
Migrate Layout components to typescript

### PR DESCRIPTION
**Pull request by fzaninotto**
Refs #4505

### CODE REVIEW

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx [hunk 1]
```diff
@@ -147,9 +149,9 @@ const useStyles = makeStyles(
     { name: 'RaReferenceArrayField' }
 );
 
-type ReferenceArrayFieldClassKey = 'progress';
+export type ReferenceArrayFieldClassKey = 'progress';
```
Comment (djhi) [inline packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx:152]: I don't get this `progress` value for the key

File: packages/ra-ui-materialui/src/field/ReferenceField.tsx [hunk 2]
```diff
@@ -237,9 +237,9 @@ ReferenceFieldView.propTypes = {
     translateChoice: PropTypes.oneOfType([PropTypes.func, PropTypes.bool]),
 };
 
-type ReferenceFieldClassKey = 'link';
+export type ReferenceFieldClassKey = 'link';
```
Comment (djhi) [inline packages/ra-ui-materialui/src/field/ReferenceField.tsx:240]: Not sure if previous comment & fix is applicable here as well

Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ReferenceField.tsx]: yes, fixed, too

File: packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx (hunk 1 above)
Comment (fzaninotto) [reply packages/ra-ui-materialui/src/field/ReferenceArrayField.tsx]: That's an old way to type the classesOverride for mui styling. Replaced with our own ClassesOverride

**Review by djhi — APPROVED** [review]

**Pull request merged**